File: crates/agent-gateway/internal/stt/volcengine_v2.go
package stt

import (
	"bytes"
	"context"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"

	"github.com/google/uuid"
	"github.com/gorilla/websocket"
	"github.com/klauspost/compress/gzip"
)

type VolcengineV2Adapter struct{}

const volcengineV2Endpoint = "wss://openspeech.bytedance.com/api/v2/asr"

func volcV2Authorization(token string) string { return "Bearer; " + token }

func (a *VolcengineV2Adapter) Test(ctx context.Context, cfg map[string]any) (string, error) {
	return readyProtocolTest(ctx, a, cfg)
}
func gzipJSON(v any) ([]byte, error) {
	raw, e := json.Marshal(v)
	if e != nil {
		return nil, e
	}
	var out bytes.Buffer
	writer := gzip.NewWriter(&out)
	if _, e = writer.Write(raw); e != nil {
		return nil, e
	}
	if e = writer.Close(); e != nil {
		return nil, e
	}
	return out.Bytes(), nil
}

func gzipBytes(payload []byte) ([]byte, error) {
	var out bytes.Buffer
	writer := gzip.NewWriter(&out)
	if _, err := writer.Write(payload); err != nil {
		return nil, err
	}
	if err := writer.Close(); err != nil {
		return nil, err
	}
	return out.Bytes(), nil
}

func volcV2Frame(messageType, flags, serialization, compression byte, payload []byte) []byte {
	frame := []byte{0x11, messageType<<4 | flags, serialization<<4 | compression, 0}
	size := make([]byte, 4)
	binary.BigEndian.PutUint32(size, uint32(len(payload)))
	return append(append(frame, size...), payload...)
}

func volcV2ResponseOK(message map[string]any) bool {
	code := int(number(message["code"]))
	text, _ := message["message"].(string)
	return code == 0 || code == 1000 || text == "Success"
}

func volcV2ResponseNoSpeech(message map[string]any) bool {
	return int(number(message["code"])) == 1013
}

func decodeVolcV2Frame(data []byte) (map[string]any, error) {
	if len(data) < 8 {
		return nil, errors.New("volcengine v2 short frame")
	}
	headerLen := int(data[0]&0x0f) * 4
	if headerLen < 4 || len(data) < headerLen+4 {
		return nil, errors.New("volcengine v2 invalid header")
	}
	messageType := data[1] >> 4
	flags := data[1] & 0x0f
	offset := headerLen
	var errorCode uint32
	var sequence int32
	if messageType == 0x0f {
		if len(data) < offset+8 {
			return nil, errors.New("volcengine v2 invalid error frame")
		}
		errorCode = binary.BigEndian.Uint32(data[offset : offset+4])
		offset += 4
	} else if messageType == 0x0b || flags&0x01 != 0 {
		if len(data) < offset+8 {
			return nil, errors.New("volcengine v2 invalid sequence frame")
		}
		sequence = int32(binary.BigEndian.Uint32(data[offset : offset+4]))
		offset += 4
	}
	payloadLen := int(binary.BigEndian.Uint32(data[offset : offset+4]))
	if payloadLen < 0 || len(data) < offset+4+payloadLen {
		return nil, errors.New("volcengine v2 invalid payload")
	}
	payload := append([]byte(nil), data[offset+4:offset+4+payloadLen]...)
	if data[2]&0x0f == 1 {
		reader, err := gzip.NewReader(bytes.NewReader(payload))
		if err != nil {
			return nil, err
		}
		payload, err = io.ReadAll(reader)
		_ = reader.Close()
		if err != nil {
			return nil, err
		}
	}
	if messageType == 0x0f {
		return map[string]any{"code": float64(errorCode), "message": string(payload)}, nil
	}
	var message map[string]any
	if err := json.Unmarshal(payload, &message); err != nil {
		return nil, err
	}
	message["_sequence"] = float64(sequence)
	message["_last"] = flags == 2 || flags == 3 || sequence < 0 || number(message["sequence"]) < 0
	return message, nil
}

func volcV2ResultText(message map[string]any) string {
	result := message["result"]
	if object, ok := result.(map[string]any); ok {
		return valueString(object, "text")
	}
	if items, ok := result.([]any); ok && len(items) > 0 {
		if object, ok := items[0].(map[string]any); ok {
			return valueString(object, "text")
		}
	}
	for _, key := range []string{"payload", "data"} {
		if wrapper, ok := message[key].(map[string]any); ok {
			if nested, ok := wrapper["result"].(map[string]any); ok {
				if text := valueString(nested, "text"); text != "" {
					return text
				}
			}
		}
	}
	return ""
}

func volcV2ResponseComplete(message map[string]any, finishing bool) bool {
	if !finishing {
		return false
	}
	lastPackage, _ := message["is_last_package"].(bool)
	lastFrame, _ := message["_last"].(bool)
	return lastPackage || lastFrame
}

func volcV2StartRequest(cfg map[string]any, id, requestID string) map[string]any {
	return map[string]any{
		"app": map[string]any{
			"appid": value(cfg, "appId"), "token": value(cfg, "accessToken"), "cluster": value(cfg, "cluster"),
		},
		"user":  map[string]any{"uid": id},
		"audio": map[string]any{"format": "raw", "rate": 16000, "bits": 16, "channel": 1, "codec": "raw"},
		"request": map[string]any{
			"reqid": requestID, "nbest": 1,
			"workflow":        "audio_in,resample,partition,vad,fe,decode,itn,nlu_punctuate",
			"show_utterances": true, "result_type": "full", "sequence": 1,
		},
	}
}
func (a *VolcengineV2Adapter) Run(ctx context.Context, id string, cfg map[string]any, commands <-chan Command, events chan<- Event) error {
	endpoint, err := websocketEndpoint(cfg, volcengineV2Endpoint)
	if err != nil {
		return stageError("VolcengineV2", "validate", err)
	}
	header := http.Header{}
	header.Set("Authorization", volcV2Authorization(value(cfg, "accessToken")))
	conn, response, err := websocket.DefaultDialer.DialContext(ctx, endpoint, header)
	if err != nil {
		return stageError("VolcengineV2", "connect", websocketConnectError(response, err))
	}
	defer func() { _ = conn.Close() }()
	requestID := uuid.NewString()
	first, err := gzipJSON(volcV2StartRequest(cfg, id, requestID))
	if err != nil {
		return stageError("VolcengineV2", "start", err)
	}
	if err = writeProviderMessage(conn, websocket.BinaryMessage, volcV2Frame(1, 0, 1, 1, first)); err != nil {
		return stageError("VolcengineV2", "start", err)
	}
	return runVolcV2Loop(ctx, conn, id, commands, events)
}
func runVolcV2Loop(ctx context.Context, conn *websocket.Conn, id string, commands <-chan Command, events chan<- Event) error {
	incoming := make(chan map[string]any, 8)
	readErr := make(chan error, 1)
	go func() {
		for {
			messageType, data, e := conn.ReadMessage()
			if e != nil {
				readErr <- stageError("VolcengineV2", "receive", e)
				return
			}
			if messageType != websocket.BinaryMessage {
				readErr <- stageError("VolcengineV2", "parse", errors.New("expected binary frame"))
				return
			}
			msg, e := decodeVolcV2Frame(data)
			if e != nil {
				readErr <- stageError("VolcengineV2", "parse", e)
				return
			}
			if !emitIncoming(ctx, incoming, msg) {
				return
			}
		}
	}()
	finishing := false
	finishSent := false
	ready := false
	pending := make([][]byte, 0, 32)
	lastText := ""
	for {
		select {
		case <-ctx.Done():
			return nil
		case e := <-readErr:
			if finishing && finishSent && isWebSocketCloseError(e, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
				if lastText != "" {
					if !emitEvent(ctx, events, Event{Type: "final", SessionID: id, Text: lastText}) {
						return ctx.Err()
					}
				}
				return nil
			}
			if finishing && finishSent {
				return stageError("VolcengineV2", "close", e)
			}
			return e
		case cmd := <-commands:
			if cmd.Cancel {
				return nil
			}
			if cmd.Audio != nil {
				if !ready {
					pending = append(pending, append([]byte(nil), cmd.Audio.PCM...))
				} else if compressed, e := gzipBytes(cmd.Audio.PCM); e != nil {
					return stageError("VolcengineV2", "send_audio", e)
				} else if e := writeProviderMessage(conn, websocket.BinaryMessage, volcV2Frame(2, 0, 0, 1, compressed)); e != nil {
					return stageError("VolcengineV2", "send_audio", e)
				}
			}
			if cmd.Finish {
				finishing = true
				if ready && !finishSent {
					compressed, compressErr := gzipBytes(nil)
					if compressErr != nil {
						return stageError("VolcengineV2", "finish", compressErr)
					}
					if e := writeProviderMessage(conn, websocket.BinaryMessage, volcV2Frame(2, 2, 0, 1, compressed)); e != nil {
						return stageError("VolcengineV2", "finish", e)
					}
					finishSent = true
				}
			}
		case msg := <-incoming:
			if volcV2ResponseNoSpeech(msg) {
				return nil
			}
			if !volcV2ResponseOK(msg) {
				return stageError("VolcengineV2", "provider_response", providerFailure("Volcengine v2", fmt.Sprint(msg["code"]), valueString(msg, "message")))
			}
			if !ready {
				ready = true
				if !emitEvent(ctx, events, Event{Type: "ready", SessionID: id}) {
					return ctx.Err()
				}
				for _, pcm := range pending {
					compressed, compressErr := gzipBytes(pcm)
					if compressErr != nil {
						return stageError("VolcengineV2", "send_audio", compressErr)
					}
					if e := writeProviderMessage(conn, websocket.BinaryMessage, volcV2Frame(2, 0, 0, 1, compressed)); e != nil {
						return stageError("VolcengineV2", "send_audio", e)
					}
				}
				pending = nil
				if finishing && !finishSent {
					compressed, compressErr := gzipBytes(nil)
					if compressErr != nil {
						return stageError("VolcengineV2", "finish", compressErr)
					}
					if e := writeProviderMessage(conn, websocket.BinaryMessage, volcV2Frame(2, 2, 0, 1, compressed)); e != nil {
						return stageError("VolcengineV2", "finish", e)
					}
					finishSent = true
				}
			}
			if text := volcV2ResultText(msg); text != "" {
				lastText = text
				if !emitEvent(ctx, events, Event{Type: "partial", SessionID: id, Text: text}) {
					return ctx.Err()
				}
			}
			if volcV2ResponseComplete(msg, finishing) {
				if lastText != "" {
					if !emitEvent(ctx, events, Event{Type: "final", SessionID: id, Text: lastText}) {
						return ctx.Err()
					}
				}
				return nil
			}
		}
	}
}

File: crates/agent-gui/src-tauri/src/services/stt/volcengine_v2.rs
use super::{
    close_provider_socket, emit, provider_failure, send_provider_message, stage_failure, text,
    websocket_endpoint, SttCommand, SttEvent,
};
use flate2::{read::GzDecoder, write::GzEncoder, Compression};
use futures_util::StreamExt;
use serde_json::Value;
use std::io::{Read, Write};
use tauri::{AppHandle, Runtime};
use tokio::sync::mpsc::Receiver;
use tokio_tungstenite::{
    connect_async,
    tungstenite::{
        client::IntoClientRequest, http::header, protocol::frame::coding::CloseCode, Message,
    },
};

fn gzip_json(value: Value) -> Result<Vec<u8>, String> {
    let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
    encoder
        .write_all(value.to_string().as_bytes())
        .map_err(|e| e.to_string())?;
    encoder.finish().map_err(|e| e.to_string())
}

fn gzip_bytes(payload: &[u8]) -> Result<Vec<u8>, String> {
    let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
    encoder.write_all(payload).map_err(|e| e.to_string())?;
    encoder.finish().map_err(|e| e.to_string())
}

fn frame(
    message_type: u8,
    flags: u8,
    serialization: u8,
    compression: u8,
    payload: &[u8],
) -> Vec<u8> {
    let mut output = vec![
        0x11,
        (message_type << 4) | flags,
        (serialization << 4) | compression,
        0,
    ];
    output.extend_from_slice(&(payload.len() as u32).to_be_bytes());
    output.extend_from_slice(payload);
    output
}

fn decode_frame(data: &[u8]) -> Result<Value, String> {
    if data.len() < 8 {
        return Err("火山 v2 帧长度无效".into());
    }
    let header_len = ((data[0] & 0x0f) as usize) * 4;
    if header_len < 4 || data.len() < header_len + 4 {
        return Err("火山 v2 帧头无效".into());
    }
    let message_type = data[1] >> 4;
    let flags = data[1] & 0x0f;
    let mut offset = header_len;
    let mut error_code = 0_u32;
    let mut sequence = 0_i32;
    if message_type == 0x0f {
        if data.len() < offset + 8 {
            return Err("火山 v2 错误帧无效".into());
        }
        error_code = u32::from_be_bytes(data[offset..offset + 4].try_into().unwrap());
        offset += 4;
    } else if message_type == 0x0b || flags & 1 != 0 {
        if data.len() < offset + 8 {
            return Err("火山 v2 序列帧无效".into());
        }
        sequence = i32::from_be_bytes(data[offset..offset + 4].try_into().unwrap());
        offset += 4;
    }
    let payload_len = u32::from_be_bytes(data[offset..offset + 4].try_into().unwrap()) as usize;
    if data.len() < offset + 4 + payload_len {
        return Err("火山 v2 负载长度无效".into());
    }
    let mut payload = data[offset + 4..offset + 4 + payload_len].to_vec();
    if data[2] & 0x0f == 1 {
        let mut decoder = GzDecoder::new(payload.as_slice());
        let mut decoded = Vec::new();
        decoder
            .read_to_end(&mut decoded)
            .map_err(|_| "火山 v2 gzip 负载无效")?;
        payload = decoded;
    }
    if message_type == 0x0f {
        return Ok(
            serde_json::json!({"code": error_code, "message": String::from_utf8_lossy(&payload)}),
        );
    }
    let mut value: Value = serde_json::from_slice(&payload).map_err(|_| "火山 v2 响应协议错误")?;
    if let Some(object) = value.as_object_mut() {
        let payload_sequence = object
            .get("sequence")
            .and_then(Value::as_i64)
            .unwrap_or_default();
        object.insert("_sequence".into(), Value::from(sequence));
        object.insert(
            "_last".into(),
            Value::from(flags == 2 || flags == 3 || sequence < 0 || payload_sequence < 0),
        );
    }
    Ok(value)
}

fn response_is_no_speech(value: &Value) -> bool {
    value.get("code").and_then(Value::as_i64) == Some(1013)
}

fn start_request(
    config: &serde_json::Map<String, Value>,
    session: &str,
    request_id: &str,
) -> Value {
    serde_json::json!({
        "app": {"appid": text(config, "appId"), "token": text(config, "accessToken"), "cluster": text(config, "cluster")},
        "user": {"uid": session},
        "audio": {"format": "raw", "rate": 16000, "bits": 16, "channel": 1, "codec": "raw"},
        "request": {
            "reqid": request_id,
            "nbest": 1,
            "workflow": "audio_in,resample,partition,vad,fe,decode,itn,nlu_punctuate",
            "show_utterances": true,
            "result_type": "full",
            "sequence": 1
        }
    })
}

pub async fn run<R: Runtime>(
    app: AppHandle<R>,
    session: String,
    config: serde_json::Map<String, Value>,
    mut rx: Receiver<SttCommand>,
) -> Result<(), String> {
    let token = text(&config, "accessToken");
    let endpoint = websocket_endpoint(&config, "wss://openspeech.bytedance.com/api/v2/asr")
        .map_err(|e| stage_failure("VolcengineV2", "validate", e))?;
    let mut request = endpoint
        .into_client_request()
        .map_err(|e| stage_failure("VolcengineV2", "connect", e.to_string()))?;
    request.headers_mut().insert(
        header::AUTHORIZATION,
        format!("Bearer; {token}")
            .parse()
            .map_err(|_| stage_failure("VolcengineV2", "connect", "Access Token 无效"))?,
    );
    let (socket, _) = connect_async(request)
        .await
        .map_err(|e| stage_failure("VolcengineV2", "connect", format!("网络错误: {e}")))?;
    let (mut write, mut read) = socket.split();
    let request_id = uuid::Uuid::new_v4().to_string();
    let start = gzip_json(start_request(&config, &session, &request_id))?;
    send_provider_message(
        &mut write,
        Message::Binary(frame(1, 0, 1, 1, &start).into()),
        "VolcengineV2",
        "start",
    )
    .await?;

    let mut ready = false;
    let mut pending = Vec::<Vec<u8>>::new();
    let mut finishing = false;
    let mut finish_sent = false;
    let mut last_text = String::new();
    loop {
        tokio::select! {
            Some(command) = rx.recv() => match command {
                SttCommand::Audio { pcm, .. } => {
                    if ready { let compressed = gzip_bytes(&pcm).map_err(|e| stage_failure("VolcengineV2", "send_audio", e))?; send_provider_message(&mut write, Message::Binary(frame(2, 0, 0, 1, &compressed).into()), "VolcengineV2", "send_audio").await?; }
                    else { pending.push(pcm); }
                }
                SttCommand::Finish => {
                    finishing = true;
                    if ready && !finish_sent {
                        let compressed = gzip_bytes(&[]).map_err(|e| stage_failure("VolcengineV2", "finish", e))?;
                        send_provider_message(&mut write, Message::Binary(frame(2, 2, 0, 1, &compressed).into()), "VolcengineV2", "finish").await?;
                        finish_sent = true;
                    }
                }
                SttCommand::Cancel => { close_provider_socket(&mut write).await; return Ok(()); }
            },
            Some(message) = read.next() => {
                let message = message.map_err(|e| stage_failure("VolcengineV2", "receive", e.to_string()))?;
                let body = match message {
                    Message::Binary(body) => body,
                    Message::Close(frame) if finishing && finish_sent && frame.as_ref().is_none_or(|frame| frame.code == CloseCode::Normal) => {
                        if !last_text.is_empty() { emit(&app, SttEvent::Final { session_id: session.clone(), text: last_text.clone() }); }
                        return Ok(());
                    }
                    Message::Close(_) => return Err(stage_failure("VolcengineV2", "close", "在结束协议前异常关闭")),
                    _ => continue,
                };
                let value = decode_frame(&body).map_err(|e| stage_failure("VolcengineV2", "parse", e))?;
                let code = value.get("code").and_then(Value::as_i64).unwrap_or(0);
                let success = code == 0 || code == 1000 || value.get("message").and_then(Value::as_str) == Some("Success");
                if response_is_no_speech(&value) {
                    close_provider_socket(&mut write).await;
                    return Ok(());
                }
                if !success { return Err(stage_failure("VolcengineV2", "provider_response", provider_failure("火山 v2", &code.to_string(), value.get("message").and_then(Value::as_str).unwrap_or_default()))); }
                if !ready {
                    ready = true;
                    emit(&app, SttEvent::Ready { session_id: session.clone() });
                    for pcm in pending.drain(..) { let compressed = gzip_bytes(&pcm).map_err(|e| stage_failure("VolcengineV2", "send_audio", e))?; send_provider_message(&mut write, Message::Binary(frame(2, 0, 0, 1, &compressed).into()), "VolcengineV2", "send_audio").await?; }
                    if finishing && !finish_sent {
                        let compressed = gzip_bytes(&[]).map_err(|e| stage_failure("VolcengineV2", "finish", e))?;
                        send_provider_message(&mut write, Message::Binary(frame(2, 2, 0, 1, &compressed).into()), "VolcengineV2", "finish").await?;
                        finish_sent = true;
                    }
                }
                let result_text = value.pointer("/result/text").and_then(Value::as_str)
                    .or_else(|| value.pointer("/result/0/text").and_then(Value::as_str))
                    .or_else(|| value.pointer("/payload/result/text").and_then(Value::as_str))
                    .or_else(|| value.pointer("/data/result/text").and_then(Value::as_str));
                if let Some(result_text) = result_text {
                    last_text = result_text.to_string();
                    emit(&app, SttEvent::Partial { session_id: session.clone(), text: last_text.clone() });
                }
                let completed = value.get("is_last_package").and_then(Value::as_bool).unwrap_or(false)
                    || value.get("_last").and_then(Value::as_bool).unwrap_or(false);
                if finishing && completed { if !last_text.is_empty() { emit(&app, SttEvent::Final { session_id: session.clone(), text: last_text.clone() }); } close_provider_socket(&mut write).await; return Ok(()); }
            }
            else => {
                if finishing && finish_sent {
                    if !last_text.is_empty() { emit(&app, SttEvent::Final { session_id: session.clone(), text: last_text.clone() }); }
                    return Ok(());
                }
                break;
            },
        }
    }
    Err(stage_failure("VolcengineV2", "close", "在结束前关闭"))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn full_request_contains_v2_auth_audio_and_workflow_fields() {
        let config = serde_json::from_value(json!({
            "appId": "app",
            "accessToken": "token",
            "cluster": "cluster"
        }))
        .expect("config map");
        let request = start_request(&config, "session", "request");
        assert_eq!(
            request.pointer("/app/token").and_then(Value::as_str),
            Some("token")
        );
        assert_eq!(
            request.pointer("/audio/format").and_then(Value::as_str),
            Some("raw")
        );
        assert_eq!(
            request.pointer("/audio/codec").and_then(Value::as_str),
            Some("raw")
        );
        assert_eq!(
            request
                .pointer("/request/show_utterances")
                .and_then(Value::as_bool),
            Some(true)
        );
        assert_eq!(
            request
                .pointer("/request/result_type")
                .and_then(Value::as_str),
            Some("full")
        );
    }

    #[test]
    fn classifies_1013_as_no_speech_only() {
        assert!(response_is_no_speech(&json!({
            "code": 1013,
            "message": "No valid speeches found in input audio"
        })));
        assert!(!response_is_no_speech(&json!({
            "code": 45000000,
            "message": "invalid request"
        })));
    }
}
